### DISCUSSION TITLE
Getting Multiple Definition error when including a utils file in multiple components

### DISCUSSION THREAD

**Discussion by epdalessandro** [nasa/fprime#2634] [Q&A] [answered]
I'm writing a F Prime project that consumes data from multiple sensors that are written in python across TCP sockets, using the built-in F' TCP client component. The parsing of the buffers in each component is very similar, so I wanted to move that functionality to a "utils.cpp" file that I included in the "Components" parent folder.

I added the utils.cpp file to the CMakeLists.txt file in the "Components" folder 
set(SOURCE_FILES 
    "${CMAKE_CURRENT_LIST_DIR}/utils.cpp"
)
and I have include guards in the utils file

```c++
#include <vector>
#include <Os/Mutex.hpp>
#ifndef UTILS
#define UTILS

class Lock {
    public:
        Lock(Os::Mutex _mutex) : mutex(_mutex) { mutex.lock(); }
        ~Lock() { mutex.unlock(); }
    private:
        Os::Mutex mutex;
};

uint64_t bytesToUint(const char* bytes, uint16_t len) { // implementation }

std::vector<uint64_t> getDataFromFields(const char* data, std::vector<uint16_t> &sizes) { // implementation }
void ProcessData(uint32_t messageSize, std::vector<char> receivedBytes, std::vector<uint16_t> &fieldSizes, void (*SubsystemDataProcessor) (uint8_t msgType)) { // implementation }

#endif
```

I'm then including this file in the `{ComponentName}.cpp` file in multiple components with #include "Components/utils.cpp"

However when I run this I get the following error:
<img width="1186" alt="Screen Shot 2024-03-28 at 1 31 16 PM" src="https://github.com/nasa/fprime/assets/71722848/000ef624-f41c-4ffd-bf2c-54abb7a5bb92">

I don't understand why there are multiple definitions if I have include guards, unless it's first linking each of the components so each component thinks that "utils.cpp" hasn't been defined, then defines it and then after each component is linked together and the compiler "finds" the multiple definitions

**Comment by LeStarch** [comment]
These are linker multiple define errors.  The issue is you've defined the functions inside a c++ file that you are treating as a header.  You should instead create `utils.hpp` and `utils.cpp` the first has definitions and the second has implementation.  Then include the `utils.hpp` in the component **not** the `utils.cpp`.

**Comment by LeStarch** [comment, marked as answer]
Here is a quick refactor that will help:

**utils.hpp**: has definitions of `class Lock` and function prototypes. 
```c++
#include <vector>
#include <Os/Mutex.hpp>
#ifndef UTILS
#define UTILS

class Lock {
    public:
        Lock(Os::Mutex _mutex) : mutex(_mutex) { mutex.lock(); }
        ~Lock() { mutex.unlock(); }
    private:
        Os::Mutex mutex;
};

uint64_t bytesToUint(const char* bytes, uint16_t len);

std::vector<uint64_t> getDataFromFields(const char* data, std::vector<uint16_t> &sizes);
void ProcessData(uint32_t messageSize, std::vector<char> receivedBytes, std::vector<uint16_t> &fieldSizes, void (*SubsystemDataProcessor) (uint8_t msgType));

#endif
```



**utils.cpp**: has implementation of functions.
```c++
#include <utils.hpp>

uint64_t bytesToUint(const char* bytes, uint16_t len) { // implementation }

std::vector<uint64_t> getDataFromFields(const char* data, std::vector<uint16_t> &sizes) { // implementation }
void ProcessData(uint32_t messageSize, std::vector<char> receivedBytes, std::vector<uint16_t> &fieldSizes, void (*SubsystemDataProcessor) (uint8_t msgType)) { // implementation }
```

**Reply by epdalessandro** [reply]
Thank you, that solved the multiple-definition problem, however now when I try to build in the deployment, it cannot find the functions in the utils file:
<img width="1110" alt="Screen Shot 2024-04-02 at 12 48 06 PM" src="https://github.com/nasa/fprime/assets/71722848/11f1c086-43cd-4cbd-858a-b09ad072ae33">

I have the utils file added to the Components CMakelists:
```
# Include project-wide components here

set(SOURCE_FILES
  "${CMAKE_CURRENT_LIST_DIR}/utils.cpp"
)

# add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/GPS/")
# add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Radio/")
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/EpsManager/")
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/AdcsManager/")
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/QuadMagManager/")
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/OpvManager/")
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/GPS/")
```

A thought we had to fix this was import the file in the CMakeLists for each component, for example:
```
set(SOURCE_FILES
  "${CMAKE_CURRENT_LIST_DIR}/../utils.cpp"
  "${CMAKE_CURRENT_LIST_DIR}/QuadMagManager.fpp"
  "${CMAKE_CURRENT_LIST_DIR}/QuadMagManager.cpp"
)
```

But this led to this error where it can't find the Fw namespace, when in the utils file I'm including the ActiveComponent Base:
```
#include "utils.hpp"
#include "Fw/Comp/ActiveComponentBase.hpp"
```

<img width="1095" alt="Screen Shot 2024-04-02 at 12 50 53 PM" src="https://github.com/nasa/fprime/assets/71722848/0d8ec83f-a40b-49cd-b088-c6406d77118d">

**Reply by epdalessandro** [reply]
The fix to this second problem was to instead create a std::function object that captures both the *this pointer and the SubsystemDataProcessor function, and pass in the NUM_MESSAGE_LENGTH_BYTES/NUM_TYPE_BYTES as parameters
